src/bootstrap/flags.rs

```diff
@@ -13,60 +13,196 @@
 //! This module implements the command-line parsing of the build system which
 //! has various flags to configure how it's run.
 
+use std::env;
 use std::fs;
 use std::path::PathBuf;
 use std::process;
-use std::slice;
 
-use getopts::Options;
+use getopts::{Matches, Options};
 
 /// Deserialized version of all flags for this compile.
 pub struct Flags {
     pub verbose: bool,
     pub stage: Option<u32>,
     pub build: String,
-    pub host: Filter,
-    pub target: Filter,
-    pub step: Vec<String>,
+    pub host: Vec<String>,
+    pub target: Vec<String>,
     pub config: Option<PathBuf>,
     pub src: Option<PathBuf>,
     pub jobs: Option<u32>,
-    pub args: Vec<String>,
-    pub clean: bool,
+    pub cmd: Subcommand,
 }
 
-pub struct Filter {
-    values: Vec<String>,
+pub enum Subcommand {
+    Build {
+        paths: Vec<PathBuf>,
+    },
+    Doc {
+        paths: Vec<PathBuf>,
+    },
+    Test {
+        paths: Vec<PathBuf>,
+        filter: Vec<String>,
+    },
+    Clean,
+    Dist {
+        install: bool,
+    },
 }
 
 impl Flags {
     pub fn parse(args: &[String]) -> Flags {
         let mut opts = Options::new();
         opts.optflag("v", "verbose", "use verbose output");
         opts.optopt("", "config", "TOML configuration file for build", "FILE");
-        opts.optmulti("", "host", "host targets to build", "HOST");
         opts.reqopt("", "build", "build target of the stage0 compiler", "BUILD");
-        opts.optmulti("", "target", "targets to build", "TARGET");
-        opts.optmulti("s", "step", "build step to execute", "STEP");
+        opts.optmulti("", "host", "host targets to build", "HOST");
+        opts.optmulti("", "target", "target targets to build", "TARGET");
         opts.optopt("", "stage", "stage to build", "N");
-        opts.optopt("", "src", "path to repo root", "DIR");
+        opts.optopt("", "src", "path to the root of the rust checkout", "DIR");
         opts.optopt("j", "jobs", "number of jobs to run in parallel", "JOBS");
-        opts.optflag("", "clean", "clean output directory");
         opts.optflag("h", "help", "print this help message");
 
-        let usage = |n| -> ! {
-            let brief = format!("Usage: rust.py [options]");
-            print!("{}", opts.usage(&brief));
+        let usage = |n, opts: &Options| -> ! {
+            let command = args.get(0).map(|s| &**s);
+            let brief = format!("Usage: x.py {} [options] [<args>...]",
+                                command.unwrap_or("<command>"));
+
+            println!("{}", opts.usage(&brief));
+            match command {
+                Some("build") => {
+                    println!("\
+Arguments:
+    This subcommand accepts a number of positional arguments of directories to
+    the crates and/or artifacts to compile. For example:
+
+        ./x.py build src/libcore
+        ./x.py build src/libproc_macro
+        ./x.py build src/libstd --stage 1 --filter hash_map
```

Comment:
Ah this is actually a way to pass extra arguments to the test harness itself (e.g. the executable). In most cases that ends up being a filter, but not always in the case of things like `-q` or such.

I'll change the name of this to `--test-args`